### PR TITLE
Bug 1186301 - [Stingray] Scanning and display BT devices in Device Deck

### PR DESCRIPTION
**Pull request by dwi2**
https://bugzilla.mozilla.org/show_bug.cgi?id=1186301

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3316d756801cd62afc294a338c4aa0dda6bcafa8

File: tv_apps/device-deck/js/bluetooth_manager.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    _keepDiscoveryHandle: function bm_keepDiscoveryHandle(handle) {
+      this._discoveryHandle = handle;
+      this._discoveryHandle.addEventListener('devicefound',
+        this._handleDeviceFound);
+    },
+
+    _startDiscovery: function bm_startDiscovery() {
+      var that = this;
+      if (!this._defaultAdapter) {
+        return Promise.reject('default adapter is not existed');
```
Comment (rexboy7) [inline tv_apps/device-deck/js/bluetooth_manager.js]: nit: `doesn't exist`

File: tv_apps/device-deck/test/unit/bluetooth_manager_test.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+      configurable: true,
+      get: function() { return targetObj; }
+    });
+  };
+
+  var setupForMockDefaultAdapter = function(context) {
+    var mockDiscoveryHandle = new MockDiscoveryHandle();
+    MockBTAdapter.state = 'enabled';
+    MockBTAdapter.startDiscovery = function() {};
+    MockBTAdapter.stopDiscovery = function() {};
```
Comment (rexboy7) [inline tv_apps/device-deck/test/unit/bluetooth_manager_test.js]: Would it be simpler if we just write
`context.sinon.stub(MockBTAdapter, stopDiscovery, function() { ... }`?

File: tv_apps/device-deck/test/unit/bluetooth_manager_test.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    var startDiscoveryStub =
+      context.sinon.stub(MockBTAdapter, 'startDiscovery', function() {
+        return Promise.resolve(mockDiscoveryHandle);
+      });
+
+    var stopDiscoverySpy =
```
Comment (rexboy7) [inline tv_apps/device-deck/test/unit/bluetooth_manager_test.js]: `stopDiscoveryStub`?

File: tv_apps/device-deck/test/unit/bluetooth_manager_test.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+      context.sinon.stub(MockBTAdapter, 'stopDiscovery', function() {
+        return Promise.resolve();
+      });
+
+    return {
+      startDiscoveryStub: startDiscoveryStub,
+      stopDiscoverySpy: stopDiscoverySpy,
+      mockDiscoveryHandle: mockDiscoveryHandle
```
Comment (rexboy7) [inline tv_apps/device-deck/test/unit/bluetooth_manager_test.js]: Although these stubs are returned and saved on another data structure (namely `mockData`
I think just call `MockBTAdapter.startDiscoveryStub.calledOnce` is more straightforward..
Do you have any considerations here?

File: tv_apps/device-deck/test/unit/bluetooth_manager_test.js [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    });
+
+    test('should invoke stopDiscovery() of defaultAdapter', function(done) {
+      bluetoothManager.safelyStopDiscovery().then(function() {
+        assert.isTrue(mockData.stopDiscoverySpy.calledOnce);
+      }).then(done, done);
+    });
+  });
+
+  suite('onDeviceFound', function() {
```
Comment (rexboy7) [inline tv_apps/device-deck/test/unit/bluetooth_manager_test.js]: nit: `onDeviceFound()`

File: tv_apps/device-deck/test/unit/bluetooth_manager_test.js (hunk 2 above)
Comment (dwi2) [inline tv_apps/device-deck/test/unit/bluetooth_manager_test.js]: Sure. I'll fix that.

File: tv_apps/device-deck/test/unit/bluetooth_manager_test.js (hunk 3 above)
Comment (dwi2) [inline tv_apps/device-deck/test/unit/bluetooth_manager_test.js]: oops, typo

File: tv_apps/device-deck/test/unit/bluetooth_manager_test.js (hunk 2 above)
Comment (dwi2) [inline tv_apps/device-deck/test/unit/bluetooth_manager_test.js]: I remember why I did this. If I don't, sinon would complains
`TypeError: Cannot stub non-existent own property stopDiscovery`

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a84951cd041aa828a395390e5bbe2f00543a5fba

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a0a670b27a7c5867afc6f64f6aab0498b0ba7e0b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=470e3529f768d15b29ffa1be5b7058586d421c76

**Pull request merged**